folio-spring-i18n/src/main/java/org/folio/spring/service/TranslationService.java

```diff
@@ -0,0 +1,347 @@
+
+  /**
+   * A map from language code -&gt; country -&gt; JSON resource containing {key -&gt; ICU format string}.
+   * Will be populated the first time any translation is requested.
+   * Note that files are not read here, only enumerated.  Look at the other maps for that information
+   *
+   * <p>This can be null - if this is the case, the map has not been constructed yet.</p>
+   * <p>You should call getFileMap instead of accessing this directly.</p>
+   *
+   * <p>Language code {@code *} will be used if one is not denoted</p>
+   */
+  @Nullable
+  protected Map<String, Map<String, TranslationFile>> translationFileFromLanguageCountryMap =
+    null;
+
+  /**
+   * A map from locale -&gt; key -&gt; ICU format string, filled in on-demand
+   * as locales are presented.
+   */
+  protected Map<Locale, TranslationMap> localeTranslations = new HashMap<>();
+
+  private final ResourcePatternResolver resourceResolver;
+  private final TranslationConfiguration configuration;
+
+  /**
+   * Get a map from language -&gt; country -&gt; pattern resource based on the classpath contents.
+   *
+   * @return the map of languages to countries to pattern resources
+   */
+  @Nonnull
+  public Map<String, Map<String, TranslationFile>> getFileMap() {
+    if (this.translationFileFromLanguageCountryMap == null) {
+      this.translationFileFromLanguageCountryMap =
+        buildLanguageCountryPatternMap();
+    }
+    return this.translationFileFromLanguageCountryMap;
+  }
+
+  /**
+   * Get the best {@link TranslationMap TranslationMap} associated with the current locale.
+   *
+   * @param locale the locale to find a TranslationMap for
+   * @param base the default translation that should be used as a fallback; can be null when
+   *   creating the default translation
+   * @return the best translation available for the current locale, potentially null if not
+   *   default is null.  It checks for match of language and country, then language, then returns
+   *   default if neither are available
+   */
+  @Nullable
+  public TranslationMap getTranslation(
+    Locale locale,
+    @Nullable TranslationMap base
+  ) {
+    return localeTranslations.computeIfAbsent(
+      locale,
+      (Locale l) -> {
+        log.info("Cache miss on " + locale);
+
+        if (this.getFileMap().containsKey(locale.getLanguage().toLowerCase())) {
+          Map<String, TranslationFile> languageMap =
+            this.getFileMap().get(locale.getLanguage().toLowerCase());
+
+          TranslationFile baseFile = languageMap.get(
+            TranslationFile.UNKNOWN_PART
+          );
+          TranslationMap baseMap = new TranslationMap(locale, baseFile, base);
+
+          if (languageMap.containsKey(locale.getCountry().toLowerCase())) {
+            return new TranslationMap(
+              locale,
+              languageMap.get(locale.getCountry().toLowerCase()),
+              baseMap
+            );
+          }
+
+          return baseMap;
+        }
```

Comment:
Within the lambda, you should use `l` instead of `locale`. References to objects outside the lambda make it into a closure, which makes it more expensive than it needs to be (if it's a closure, then it needs to be instantiated in order to store the reference before the lambda is called. If it's a pure function, then it can be a static method)